internal/controllers/topology/cluster/desired_state.go

```diff
@@ -623,10 +623,19 @@ func computeMachineDeployment(_ context.Context, s *scope.Scope, desiredControlP
 		},
 	}
 
-	// If it's a new MachineDeployment, set the finalizer.
-	// Note: we only add it on creation to avoid race conditions later on when
-	// the MachineDeployment topology controller removes the finalizer.
-	if currentMachineDeployment == nil {
+	// Set the finalizer on the MachineDeployment.
+	// Note - setting this in every case except where the MachineDeployment is in deleting AND the finalizer has already been removed
+	// prevents the finalizer from being removed on server-side apply as it is not explicitly set.
+	// The MachineDeployment topology controller removes the finalizer.
+
+	// Set the finalizer:
+	// - If this is a new MachineDeployment.
+	if currentMachineDeployment == nil ||
+		// - If the MachineDeployment is not being deleted.
+		currentMachineDeployment.Object.DeletionTimestamp.IsZero() ||
+		// - If the MachineDeployment is being deleted but currently still has the Finalizer.
+		(!currentMachineDeployment.Object.DeletionTimestamp.IsZero() &&
+			controllerutil.ContainsFinalizer(currentMachineDeployment.Object, clusterv1.MachineDeploymentTopologyFinalizer)) {
```

Comment:
my only concern is a race between Cluster Topology controller holding the finalizer and the MachineDeployment topology controller removing it
is this a valid concern? if yes, there is a couple of ways to get around this...

Reply:
It should be fine as even if we have a race it won't occur permanently.

Also the MD is usually deleted by the Cluster topology controller and in this case we only do a Delete on the MD